util: add isArrayBufferDetached method (#45512)
nodejs/node: doc/api/util.md

The API is accessible via `require('node:util').types` or `require('node:util/types')`.

> ### `util.types.isArrayBufferDetached(value)`

Reviewer: a) I think this list is alphabetical

b) Should this really be a public API? Why is checking whether `new TypedArray(buffer)` throws insufficient?

Author: I fixed the alphabetical issue.

<!-- YAML
added: REPLACEME
-->

* `value` {any}
* Returns: {boolean}

Returns `true` if the value is a built-in [`ArrayBuffer`][] and
is detached.

```js
util.types.isArrayBufferDetached(null); // Returns false
util.types.isArrayBufferDetached(new ArrayBuffer()); // Returns false

> ```

Reviewer 2: doh, github glitched, sigh... I guess the Pr was updated while my review here was still pending... lol. This should be down with the actual isArrayBufferDetached example

> ```

Author: Thanks! I just updated the documentation. Sorry for the confusion!

util.types.isAnyArrayBuffer(new SharedArrayBuffer());  // Returns true
```

### `util.types.isArrayBufferDetached(value)`

<!-- YAML
added: REPLACEME
-->

* `value` {any}
* Returns: {boolean}

Returns `true` if the value is a built-in [`ArrayBuffer`][] and

> is detached.

Reviewer 2: A comment here quickly explaining what "detached" means might be helpful to newcomers.

Author: Thanks for the review. I updated the pull request.

util.types.isAnyArrayBuffer(new SharedArrayBuffer());  // Returns true
```

### `util.types.isArrayBufferDetached(value)`

<!-- YAML
added: REPLACEME
-->

* `value` {any}
* Returns: {boolean}

Returns `true` if the value is a built-in [`ArrayBuffer`][] and

> is detached. Detached arrays have a byte length of 0, which prevents
> JavaScript from ever accessing underlying backing store.
> 
> For example, we can end up with a detached buffer when using a BYOB
> (bring your own buffer) on a ReadableStream.

Reviewer 3: ```suggestion
is detached. For example, we can end up with a detached buffer
when using a BYOB (bring your own buffer) on a ReadableStream.
Detached arrays have a `byteLength` of 0, and their contents can
not be accessed in JavaScript.
```
The original text sounds a bit misleading: it's not length of 0 that prevents us from using detached buffers but detached state itself.

Author: Thank you for the review. I updated the pull request.

```js

> util.types.isArrayBufferDetached(null); // Returns false
> util.types.isArrayBufferDetached(new ArrayBuffer()); // Returns false

Reviewer 3: Would be nice to have a code example that returns true.
```suggestion
util.types.isArrayBufferDetached(null); // Returns false
util.types.isArrayBufferDetached(new ArrayBuffer()); // Returns false

const { buffer } = new Uint8Array([1, 2, 3]);
new MessageChannel().port1.postMessage('', [buffer]);
util.types.isArrayBufferDetached(buffer); // Returns true
```

Author: Thank you for the review. I updated the pull request.
